fs: improve argument handling for ReadStream

ryzokuken opened the pull request:
Improve handling of erratic arguments in fs.ReadStream

Refs: https://github.com/nodejs/node/pull/19732

For the uninitiated, a summary of what I'm planning to change:

| Type | Current Behavior | Proposed Behavior |
|---|---|---|
| `undefined` | (0, `Infinity`) | (0, `Infinity`) |
| Anything other than `Number`   | `ERR_INVALID_ARG_TYPE` | `ERR_INVALID_ARG_TYPE` |
| `NaN` | `ERR_INVALID_ARG_TYPE` | `ERR_OUT_OF_RANGE` |
| Negative `Number`s | ✘ | `ERR_OUT_OF_RANGE` |
| Fractional `Number`s | ✘ | `Math.round` |
| Integers | Work Perfectly | Still work perfectly |
| Invalid Arguments | `ERR_OUT_OF_RANGE` | `ERR_OUT_OF_RANGE` |

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

cc @addaleax @anliting @nodejs/fs

ryzokuken commented:
P.S. Don't run the tests yet, they'd probably need to be changed as this should *probably* be `semver-major`, as we're now throwing on values we previously accepted.

Just take a look at the code and let me know what you think about it (for now).

thefourtheye commented on lib/fs.js:
```diff
@@ -2008,31 +2008,55 @@ function ReadStream(path, options) {
+
+  // Case 4: If either start or end is fractional, round them to the nearest
+  // whole number. (Not integer as negatives not allowed)
+  if (!Number.isInteger(this.start)) {
```
Now that we are doing much stricter checks, `isSafeInteger` would be better I guess.

BridgeAR replied:
@thefourtheye I think that does not apply here.

anliting replied:
@thefourtheye I think that does not too.

tniessen commented:
Is there a reason to make `NaN` be `ERR_INVALID_ARG_TYPE` instead of `ERR_OUT_OF_RANGE`? I would expect "Invalid type" if a type mismatches according to `typeof` or according to class / interface expectations, both don't apply here (`typeof NaN === 'number'`). From a practical point of view, `NaN` is most probably the result of an incorrect computation (e.g. `Math.sqrt(-1)`), and in such cases `ERR_OUT_OF_RANGE` might be easier to understand.

BridgeAR commented:
@tniessen I think you misread the change. As far as I see it, your suggestion is already the case in this PR.

ryzokuken commented:
@tniessen @BridgeAR exactly. `ERR_INVALID_ARG_TYPE` is the error currently being thrown, and this PR switches to `ERR_OUT_OF_RANGE`, completely incorporating your suggestion.

ryzokuken commented:
@ronkorving I had something similar in mind, but Anna convinced me otherwise and with good reason. A lot of people do a ton of math to get the values of `start` and `end`, and things might not go exactly as planned.

Quoting @addaleax here:

> When your math gives you 12.9999998, which should have turned out to 13 if it weren’t for floating-point inaccuracies, and we round down to 12, then all your offsets into the returned buffers are wrong.

For a better insight into our conversations and how we came to this final decision, read the thread at the reference PR (https://github.com/nodejs/node/pull/19732).

anliting commented:
@ryzokuken 

I haven't reviewed the code because I am a little busy these days.

Would you like to write/quote the reason of using `Math.round` etc. in the code? I used to do that in my own projects to prevent good people from wasting creative mind on checking the code.

anliting commented:
I am wondering if `(-Infinity,...)` or `(Infinity,Infinity)` breaks the method......

ryzokuken commented:
@anliting No idea, we could test for that, though. I'll be adding/updating unit tests for `ReadStream` today.

anliting commented:
I mean, you program seems to allow passing `Infinity` for `start`, but I am not sure if the other parts will break on `this.start==Infinity`.

ryzokuken commented:
Added new tests and updated existing ones, take a look when you can.

The tests might look a little too much, but better be safe than be sorry, I guess.

addaleax commented on lib/fs.js:
```diff
@@ -2008,31 +2008,55 @@ function ReadStream(path, options) {
-    }
+  // Case 0: If either start or end is undefined, set them to defaults
+  // (0, Infinity) respectively.
+  this.start = this.start === undefined ? 0 : this.start;
```
I’m mostly sure this is the cause of the test failure – `start: 0` and `start: undefined` aren’t equivalent, because the former means that the stream will use reads with an `offset` parameter (which doesn’t work on all types of files). If `start: undefined` was passed, this value should be `undefined`, too.

ryzokuken replied:
Alright, let me just remove the condition then.

ryzokuken commented:
@addaleax better now?

addaleax commented on lib/fs.js:
```diff
@@ -2008,31 +2008,53 @@ function ReadStream(path, options) {
-    }
+  // Case 1: If the type of either start or end is not number, throw
+  // ERR_INVALID_ARG_TYPE (TypeError).
+  if (typeof this.start !== 'number') {
```
I think you want to allow `undefined` explicitly here and for the other checks as well?

(The `if (this.start !== undefined) {` block from the original variant of this code seems like a pretty reasonable idea, tbh…)

ryzokuken replied:
Okay. So, should I enclose *all* the type-checking code within that statement?

addaleax replied:
@ryzokuken All that’s related to `start`, yes. `end` is another story, and my original `TODO` was more or less about keeping the `end` checks *out* of such a block :)

ryzokuken replied:
Exactly. That's why I wanted to keep the checks separate, so they worked either way.

I've enclosed all this code into the `if` block, which *should* work if `end` does not need to be transformed or checked at all if `start` is `undefined`. That probably isn't the case, is it?

ryzokuken commented:
Umm, okay. So a couple of other `ReadStream` related tests `throw` now.

The arguments passed don't qualify by the newer standards, I guess.

ryzokuken commented:
@addaleax How about this instead?

addaleax commented on lib/fs.js:
```diff
@@ -2009,15 +2009,34 @@ function ReadStream(path, options) {
+    }
+    if (!Number.isInteger(this.end)) {
+      this.end = Math.round(this.end);
+    }
```
I guess we would want to do the type checking for `end` before we get to this block? Otherwise `end: undefined` would get converted to `Math.round(undefined)` here, which is `NaN`… if I read the code correctly?

ryzokuken replied:
Holy moly. I missed this one and got a few failing tests for `NaN`. Stupid me, I guess. Thanks for pointing this out.

addaleax commented on lib/fs.js:
```diff
@@ -2009,15 +2009,34 @@ function ReadStream(path, options) {
     }
 
+    // Case 4: If start is fractional, round them to the nearest whole number.
+    // (Not integer as negatives not allowed)
```
I’m a bit confused by the comment here – we do round to the nearest integer, right?

addaleax commented on lib/fs.js:
```diff
@@ -2009,15 +2009,31 @@ function ReadStream(path, options) {
+    // Case 5: If start is a whole number, work perfectly.
+
+    // Case 6: If start is greater than end, throw ERR_OUT_OF_RANGE (RangeError)
     if (this.start > this.end) {
```
I think this works fine, but as far as I can tell that is because `end: undefined` gets coerced to `NaN` here (which makes this condition always false), and that the fact that we later round `end` does not affect this comparison

Maybe it would be a bit more obvious that this works if we swap type checking for `end` and `start`?

addaleax replied:
@ryzokuken What I meant was, `end` is being coerced to `NaN` in this comparison if it was passed in as `undefined`, because we haven’t yet gotten to the part where it’s converted to `Infinity`.

ryzokuken replied:
Oh. so, should I move the block that handles `this.end === undefined` first and move to a similar type-check below?

Let me make it and demonstrate.

ryzokuken commented:
Okay, looks like the tests passed for me this time around, so that's probably a good sign.

ryzokuken commented:
@addaleax it does feel much better now, doesn't it? Definitely more consistent (`start !== undefined` vs `end !== undefined`).

anliting commented:
@ryzokuken It seems that an `undefined` value of `start` is not set to `0`. Is that intentional?

anliting commented on lib/fs.js:
```diff
@@ -2042,10 +2045,7 @@ function ReadStream(path, options) {
-  if (this.end === undefined) {
-    this.end = Infinity;
-  } else {
+  if (this.end !== undefined) {
```
It seems that `this.end` will never be `undefined` here.

ryzokuken replied:
It won't. I realize what's wrong here, will change this clause. Thanks for pointing out.

ryzokuken commented:
> @ryzokuken It seems that an undefined value of start is not set to 0. Is that intentional?

@anliting Yes, that was intentional. Setting `start` to `0` on `undefined` breaks the current functionality which I don't plan to do in this PR.

ryzokuken commented:
Okay, so @anliting raised four points:

- An undefined value of this.start is not set to 0.

This is intentional. See https://github.com/nodejs/node/pull/19898#issuecomment-380320611

- Doesn't allow -0.1 to be rounded. Commenting it makes it fine, but my program allows it.

Again, it was intentional on my part.

- "if (this.end !== undefined){" on L2048 is useless because of "this.end = this.end === undefined ? Infinity : this.end;" on L2012.

Quite true. I'm thankful to you for point this out. Not sure how I missed it.

- "this.start > this.end" is checked before this.end is rounded.

I fail to see how this is a big issue, but it is an issue nonetheless, and I fixed it. Thank you again.

anliting commented:
@ryzokuken: "I fail to see how this is a big issue, but it is an issue nonetheless, and I fixed it. Thank you again."

Not really big, just something like `{start:1, end: 0.99}`.

ryzokuken commented:
@addaleax this one has been open for over 48 hours now and had no criticisms/reviews in the last ~16 hours. Could we land this?

addaleax commented:
@ryzokuken There are some good points mentioned by @anliting. Switching the validation for `start` and `end` so that `end` comes first would resolve at least one of them, so I think we should do that?

Also, re: landing … yes, we can do that, but keep in mind that this is a `semver-major` PR, so it cannot enter an actual release until October. It’s a good idea to land it soon-ish, but I (personally) try to keep my own `semver-major` PRs open for a while longer than 2 days…

ryzokuken commented:
@addaleax I pushed a commit which addressed pretty much all their concerns.

re: `semver-major`, I understand. Let's wait a little more, then.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/14225/

ryzokuken commented:
@mcollina @addaleax had asked me to do so in the initial PR (https://github.com/nodejs/node/pull/19732) and we had decided to add documentation in a subsequent follow-up PR.

If you insist, I'd add it in this one though, I'm okay with either. (but would prefer 2 PRs over 1, keeps things simple IMO)

ryzokuken commented:
Also, I had been waiting for people to agree on the specifics of the behavior of this function (which *did* change over and over again, so I'm glad), which is also why 2 PRs would make more sense.

If you approve the current behavior, then I could be more confident with the current behavior and document it.

BridgeAR commented:
@ryzokuken behavior changes should always be documented in the PR that lands the change as far as I know.

ryzokuken commented:
@BridgeAR in that case, I'll change the docs promptly.

ryzokuken commented:
@mcollina @BridgeAR how does this look?

ryzokuken commented:
If you're okay with the code changes, should I squash all those commits into one?

I'm thinking of keeping three - one for tests, one of the code and another for docs.

addaleax commented on doc/api/fs.md:
```diff
@@ -1313,6 +1313,19 @@ start counting at 0. If `fd` is specified and `start` is omitted or `undefined`,
+- Both `start` and `end` should be integer values. If they are not, then they
+are rounded to the closest integer using `Math.round`.
+- If none of the above rules apply, then the function works without any quirks
+as expected.
```
I’d move this closer to the end of the documentation part, since it’s mostly writing out facts that are more or less implicit in the `start` and `end ` descriptions already.

(Maybe @mcollina was rather talking about adding a `changes:` block to the metadata? That would make more sense to me than listing these conditions, tbh …)

ryzokuken replied:
In that case, I must've misunderstood them. Awaiting their response, if this is not what they asked for, I'd make the required changes and remove this.

mcollina replied:
I just generically said that this needed some documentation as it is a semver-major change.

A changes: block in the metadata should be added. Also, I would follow @addaleax suggestion.

I’d review this later (might be next week).

ryzokuken replied:
Oh, from metadata you meant something like:

```yaml
<!-- YAML
added: v0.1.90
changes:
  - version: v8.8.0
    pr-url: https://github.com/nodejs/node/pull/15380
    description: The `windowsHide` option is supported now.
-->
```

right? Will make a commit doing that within the next few hours.

ryzokuken replied:
@addaleax which version would this be landing in, `v10.0.0`?

addaleax replied:
@ryzokuken The magic word for version numbers is `REPLACEME`. It’s going to land in `v11.0.0`, but please always use the former :)

ryzokuken replied:
TIL. Will make the change.

BridgeAR replied:
Some comments about the added documentation: 

> If `end` is omitted or set to undefined, then it will be coerced to `Infinity`.

Instead, please just write: `**Default**: Infinity` next to the argument.

> The type of both `start` and `end` needs to be `number`.

That part is obsolete because the type is already declared at the function header.

> If not, then the function would throw a `TypeError` with the code `ERR_INVALID_ARG_TYPE`. Both `start` and `end` should be valid values (i.e. they should neither be negative nor `NaN`, and `start` should not be greater than `end`). If they are, the function would throw a `RangeError` with the code `ERR_OUT_OF_RANGE`.

That should not be documented. It would be to verbose to document these things everywhere and it would be outdated super fast. Especially as most of it should be clear from the function header.

> Both `start` and `end` should be integer values. If they are not, then they are rounded to the closest integer using `Math.round`.

This can be shortened to: `If 'start' or 'end' are not integers, then they are rounded to the closest integer using 'Math.round'.`

> If none of the above rules apply, then the function works without any quirks as expected.

That should not be documented as that should always be the case ;-)

ryzokuken replied:
@BridgeAR quite right. Will remove the stuff as mentioned.

ryzokuken commented:
@addaleax @mcollina I suppose this is better? Let me know if something's amiss.

BridgeAR commented on lib/fs.js:
```diff
@@ -2009,30 +2009,67 @@ function ReadStream(path, options) {
   if (this.start !== undefined) {
-    if (typeof this.start !== 'number' || Number.isNaN(this.start)) {
+    // Case 1: If the type of start is not 'number', throw ERR_INVALID_ARG_TYPE
+    // (TypeError).
```
To me these comments seem redundant. It should be clear from the code what happens. So I personally would remove all added comments. What do others think about that?

ryzokuken replied:
@BridgeAR sure, will do.

ryzokuken commented:
@BridgeAR left only what was totally necessary. Take a look now if it looks better.

BridgeAR commented on doc/api/fs.md:
````diff
@@ -1298,6 +1303,7 @@ default value of 64 kb for the same parameter.
 
 ```js
 const defaults = {
+  end: Infinity,
````
@ryzokuken `start` is set to `0` by default if I am not mistaken (implicit).

@vsemozhetbyt seems like the documentation of `fs` is not aligned with most other documentations when it comes to e.g. default values and return types. Is there an issue to track and improve this?

addaleax replied:
@BridgeAR The default for `start` is `undefined`, i.e. reading from the current position in the file

ryzokuken replied:
@BridgeAR exactly as @addaleax said. If `start` is unset, we're not setting it to any default value.

vsemozhetbyt replied:
@BridgeAR I will try.

ryzokuken replied:
@vsemozhetbyt I could help if you need any.

vsemozhetbyt replied:
I will post a PR link here, you can review)

But there may be conflicts with this PR as I plan to incorporate objects with defaults into parameter lists an delete the objects. You or I will need to rebase)

vsemozhetbyt replied:
@BridgeAR, @ryzokuken, PTAL: https://github.com/nodejs/node/pull/20050

vsemozhetbyt commented on doc/api/fs.md:
```diff
@@ -1313,6 +1319,9 @@ start counting at 0. If `fd` is specified and `start` is omitted or `undefined`,
 The `encoding` can be any one of those accepted by [`Buffer`][].
 
+If `start` or `end` are not integers, then they are rounded to the closest
+integer using `Math.round`.
```
Nit: `` `Math.round` `` -> `` `Math.round()` ``?

ryzokuken replied:
Will do, gimme a sec.

vsemozhetbyt commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/14306/

BridgeAR commented:
@addaleax I just checked the rest of `fs` and we actually have the tendency of strictly testing for integers (`validateLen`, `validateUint32`). I would really like to prevent a mixture of validations by introducing rounding the input. As far as I know, that was never the case anywhere else before.

I would like to have a consistent way of dealing with this, so adding the `blocked` label for now until it is decided what to do.

ryzokuken commented:
@BridgeAR see https://github.com/nodejs/node/pull/19732#issuecomment-379499985 and https://github.com/nodejs/node/pull/19732#issuecomment-379500305

vsemozhetbyt commented:
Sorry for the conflict)

ryzokuken commented:
@vsemozhetbyt no worries. Will rebase.

ryzokuken commented:
@vsemozhetbyt @BridgeAR @mcollina @addaleax so, I guess all concerns have been taken care of?

ryzokuken commented:
@BridgeAR awaiting your approval before landing this.

ryzokuken commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14371/

devsnek commented:
@jasnell what's the status on this being able to be landed with regard to 10.0.0?

boneskull commented:
While I agree rounding fractional numbers is the Right Thing to do (as per the discussion in #19732), unless we can show that the current behavior is causing an Actual Person to have Actual Problems, I'm not convinced a breaking change to how they are handled is warranted; the change itself could cause Actual Problems for Actual People where none existed before.  :smile:

ryzokuken commented:
@boneskull accepting negative numbers sounds irrational at best, and I'd think you'd agree with it.

Plus, this PR doesn't only start throwing at negative numbers, it makes error checking tighter and error a tad more descriptive.

Also, it decouples the sanity checks for `end` and `start`, addressing the TODO at:

https://github.com/nodejs/node/blob/94596560c2556fae614ef237ed0ff2f749b14651/lib/fs.js#L2034-L2036

Let me know if anything still sounds amiss.

ryzokuken commented:
P.S. I personally think anyone who'd break their code after this one was probably doing it wrong in the first place, but I'd love to have your thoughts on that.

boneskull commented:
The rest of the PR sounds fine to me, just not the rounding bit.

ryzokuken commented:
I'd love to discuss it. The behavior of rounding numbers to the closest integer sounds unpredictable to me too, and I'd rather predictably round it to, say, the immediately smaller integer (`Math.floor`).

Perhaps we could reach consensus on an alternate behavior instead?

boneskull commented:
"Sounding unpredictable" just isn't good justification for potentially breaking "production code, working as intended".  If we had a non-contrived, real-world example of where the current behavior's unpredictability is causing problems, then that'd be good justification enough.

ryzokuken commented:
I think you got me wrong. I agreed that the behavior I just added (one proposed in the PR) is unpredictable, and invited you to discuss alternatives.

devsnek commented:
my thought process is basically that if we want users to pass integers, we should make them pass integers. lets not be shy about telling them when they're using an api incorrectly.

ryzokuken commented:
@BridgeAR @devsnek I hope this definitely looks better?

ryzokuken commented on lib/fs.js:
```diff
@@ -2014,30 +2014,49 @@ function ReadStream(path, options) {
-    } else if (typeof this.end !== 'number' || Number.isNaN(this.end)) {
-      throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+
+    if (Number.isNaN(this.start)) {
```
I would rather throw a little more descriptively, let the user know what went wrong.

That said, I'd love to hear everyone's opinions on this, and would love to change this accordingly.

BridgeAR replied:
There are a couple PRs that consolidate validation functions. It is possible to improve the check here performance and code wise but functionality wise it is fine and I believe it is best to land it as is. It can still be improved later on.

ryzokuken replied:
But if you think about it, `NaN` and floats aren't the wrong type, are they? They're still `Number`s.

They are invalid arguments because they lie outside of the rational input range.

ronkorving replied:
I didn't realize the type was only allowed to be a JS type, not a type in the broader sense (like integer).

ryzokuken commented on doc/api/fs.md:
```diff
@@ -1325,6 +1330,9 @@ start counting at 0. If `fd` is specified and `start` is omitted or `undefined`,
 `fs.createReadStream()` reads sequentially from the current file position.
 The `encoding` can be any one of those accepted by [`Buffer`][].
 
+If `start` or `end` are not integers, then they are rounded to the closest
```
Will make this change, thanks.

Sorry to have missed it.

ryzokuken commented:
@ronkorving @BridgeAR I think it's all good now, right?

BridgeAR commented:
LGTM

ryzokuken commented:
@BridgeAR how do I proceed with this? Should I submit this one to CITGM or something?

I don't believe we could just normally "land" `semver-major` PRs, can we?

apapirovski commented:
Now that it's had significant changes, since it's `semver-major`, it should get 2 re-approvals from @nodejs/tsc before it lands. But other than that, yes you can just land it once it has all the appropriate sign offs and CI + CitGM runs.

ryzokuken commented:
Final (hopefully) CI: https://ci.nodejs.org/job/node-test-pull-request/14477/

ryzokuken commented:
I fixed the error message (it wasn't quite right), but the exact code is:

```js
if (this.end === undefined) {
    this.end = Infinity;
  } else {
...
    if (!Number.isInteger(this.end)) {
      throw new ERR_OUT_OF_RANGE('end', 'integer', this.end);
    }
...
  }
```

It fails because `end` is `Infinity`, i.e. at https://github.com/ryzokuken/node/blob/4863dbac9ee238e14e4a197e6db0fdeec2c5b417/test/parallel/test-fs-read-stream-throw-type-error.js#L40

should I change the condition such that it's:

```patch
diff --git a/lib/fs.js b/lib/fs.js
index db7ab21b7c..8f698d0229 100644
--- a/lib/fs.js
+++ b/lib/fs.js
@@ -2033,7 +2033,7 @@ function ReadStream(path, options) {
     this.pos = this.start;
   }
 
-  if (this.end === undefined) {
+  if (this.end === undefined || this.end === Infinity) {
     this.end = Infinity;
   } else {
     if (typeof this.end !== 'number') {
```

That *should* fix the failures, I guess.

ryzokuken commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14497/

apapirovski commented on lib/fs.js:
```diff
@@ -2014,30 +2014,49 @@ function ReadStream(path, options) {
-  // TODO(addaleax): Make the above typecheck not depend on `start` instead.
-  // (That is a semver-major change).
-  if (typeof this.end !== 'number')
+  if (this.end === undefined || this.end === Infinity) {
```
You could save an unnecessary assignment if you move this condition to the else and make it `else if (this.end !== Infinity)`.

ryzokuken replied:
Will do, sure.

ryzokuken commented:
Rerunning CI: https://ci.nodejs.org/job/node-test-pull-request/14511/

ryzokuken commented:
Doesn't look like a JS test failure. Someone please take a look?

ryzokuken commented:
https://ci.nodejs.org/job/node-test-commit-freebsd/nodes=freebsd11-x64/17308/consoleFull failed with a Java error:

```console
23:03:54 FATAL: command execution failed
23:03:54 java.nio.channels.ClosedChannelException
23:03:54 	at org.jenkinsci.remoting.protocol.impl.ChannelApplicationLayer.onReadClosed(ChannelApplicationLayer.java:208)
23:03:54 	at org.jenkinsci.remoting.protocol.ApplicationLayer.onRecvClosed(ApplicationLayer.java:222)
23:03:54 	at org.jenkinsci.remoting.protocol.ProtocolStack$Ptr.onRecvClosed(ProtocolStack.java:832)
23:03:54 	at org.jenkinsci.remoting.protocol.FilterLayer.onRecvClosed(FilterLayer.java:287)
23:03:54 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.onRecvClosed(SSLEngineFilterLayer.java:181)
23:03:54 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.switchToNoSecure(SSLEngineFilterLayer.java:283)
23:03:54 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.processWrite(SSLEngineFilterLayer.java:503)
23:03:54 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.processQueuedWrites(SSLEngineFilterLayer.java:248)
23:03:54 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.doSend(SSLEngineFilterLayer.java:200)
23:03:54 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.doCloseSend(SSLEngineFilterLayer.java:213)
23:03:54 	at org.jenkinsci.remoting.protocol.ProtocolStack$Ptr.doCloseSend(ProtocolStack.java:800)
23:03:54 	at org.jenkinsci.remoting.protocol.ApplicationLayer.doCloseWrite(ApplicationLayer.java:173)
23:03:54 	at org.jenkinsci.remoting.protocol.impl.ChannelApplicationLayer$ByteBufferCommandTransport.closeWrite(ChannelApplicationLayer.java:313)
23:03:54 	at hudson.remoting.Channel.close(Channel.java:1446)
23:03:54 	at hudson.remoting.Channel.close(Channel.java:1399)
23:03:54 	at hudson.slaves.SlaveComputer.closeChannel(SlaveComputer.java:746)
23:03:54 	at hudson.slaves.SlaveComputer.access$800(SlaveComputer.java:99)
23:03:54 	at hudson.slaves.SlaveComputer$3.run(SlaveComputer.java:664)
23:03:54 	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
23:03:54 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
23:03:54 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
23:03:54 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
23:03:54 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
23:03:54 	at java.lang.Thread.run(Thread.java:748)
23:03:54 Caused: java.io.IOException: Backing channel 'JNLP4-connect connection from 107.170.28.213/107.170.28.213:25810' is disconnected.
23:03:54 	at hudson.remoting.RemoteInvocationHandler.channelOrFail(RemoteInvocationHandler.java:209)
23:03:54 	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:278)
23:03:54 	at com.sun.proxy.$Proxy86.isAlive(Unknown Source)
23:03:54 	at hudson.Launcher$RemoteLauncher$ProcImpl.isAlive(Launcher.java:1137)
23:03:54 	at hudson.Launcher$RemoteLauncher$ProcImpl.join(Launcher.java:1129)
23:03:54 	at hudson.tasks.CommandInterpreter.join(CommandInterpreter.java:155)
23:03:54 	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:109)
23:03:54 	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:66)
23:03:54 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
23:03:54 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:744)
23:03:54 	at hudson.model.Build$BuildExecution.build(Build.java:206)
23:03:54 	at hudson.model.Build$BuildExecution.doRun(Build.java:163)
23:03:54 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:504)
23:03:54 	at hudson.model.Run.execute(Run.java:1727)
23:03:54 	at hudson.matrix.MatrixRun.run(MatrixRun.java:153)
23:03:54 	at hudson.model.ResourceController.execute(ResourceController.java:97)
23:03:54 	at hudson.model.Executor.run(Executor.java:429)
```

tniessen commented:
@ryzokuken That's an infrastructure failure, test results look good. You can try rerunning specific configurations in such cases.

ryzokuken commented:
Rebuilding PLinux: https://ci.nodejs.org/job/node-test-commit-plinux/17237/

TimothyGu commented on lib/fs.js:
```diff
@@ -2014,30 +2014,49 @@ function ReadStream(path, options) {
+      throw new ERR_OUT_OF_RANGE('start', 'positive', this.start);
+    }
+
+    if (!Number.isInteger(this.start)) {
```
I would instead use `Number.isSafeInteger`. `this.start` may be unreliable if it is higher than `Number.MAX_SAFE_INTEGER`.

ryzokuken replied:
Agreed. Made the change.

TimothyGu commented on lib/fs.js:
```diff
@@ -2014,30 +2014,49 @@ function ReadStream(path, options) {
+      throw new ERR_OUT_OF_RANGE('end', 'positive', this.end);
+    }
+
+    if (!Number.isInteger(this.end)) {
```
Ditto.

ryzokuken commented:
@TimothyGu done.

BridgeAR commented on lib/fs.js:
```diff
@@ -2014,30 +2014,49 @@ function ReadStream(path, options) {
+    }
+
+    if (!Number.isSafeInteger(this.end)) {
+      throw new ERR_OUT_OF_RANGE('end', 'integer', this.end);
```
This is now not specific enough anymore. E.g., `2 ** 54` would now report it is not an integer. The same above.

ryzokuken commented:
@BridgeAR this must look good, right?

apapirovski commented:
@jasnell @addaleax @BridgeAR @ronkorving PTAL. This has undergone significant changes and should be reviewed again before landing.

BridgeAR commented on lib/fs.js:
```diff
@@ -2014,30 +2014,49 @@ function ReadStream(path, options) {
+    }
+
+    if (!Number.isSafeInteger(this.end)) {
+      throw new ERR_OUT_OF_RANGE('end', 'safe integer', this.end);
```
The message will be: `The value of "end" is out of range. It must be safe integer. Received XXX`.
That should be further improved. My suggestion is to do this in two steps:

One would say: `It must be an integer`. If it is not a safe integer, I would say: `It must be => 0 and <= 2 ** 53 - 1`.

So the code block here should look like e.g.:

```js
if (!Number.isSafeInteger(this.end)) {
  if (typeof this.end !== 'number')
    throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
  if (!Number.isInteger(this.end))
    throw new ERR_OUT_OF_RANGE('end', 'an integer', this.end);
  throw new ERR_OUT_OF_RANGE('end', '=> 0 and <= 2 ** 53 - 1', this.end);
} else if (this.end < 0) {
  throw new ERR_OUT_OF_RANGE('end', => 0 and <= 2 ** 53 - 1, this.end);
}
if (this.start !== undefined && this.start > this.end) {
  // Code below
}
```

The reason for this code is: a) it is faster because it needs less code path and less checks. b) it unifies the checks to give a consistent error message. c) the error message is clear (NaN will not be >= 0 or <= 2 ** 53 - 1).

To actually realize that the error message is fine or not, it would be nice to also check for the message in the tests.
To do that, I would actually use a loop with values that I want to check for that will all result with the same error message as in:

```js
[
  val1,
  val2,
  val3
].forEach((options) => {
  createReadStreamErr(example, options, rangeError);
});
```

ryzokuken replied:
@BridgeAR sounds good. Will make changes as required.

ryzokuken replied:
> `throw new ERR_OUT_OF_RANGE('end', '=> 0 and <= 2 ** 53 - 1', this.end);`

I think you mean `>=`, right? Just wanted to confirm.

BridgeAR commented on lib/fs.js:
```diff
@@ -2014,30 +2014,49 @@ function ReadStream(path, options) {
+
+    if (this.start !== undefined && this.start > this.end) {
+      const errVal = `{start: ${this.start}, end: ${this.end}}`;
+      throw new ERR_OUT_OF_RANGE('start', '<= "end"', errVal);
```
Looking at this again I would also change this to:

```js
throw new ERR_OUT_OF_RANGE('start', `<= ${this.end}`, this.start);
```

This ends up as:

`The value of "start" is out of range. It must be <= ${this.end}. Received ${this.start}`

instead of:

`The value of "start" is out of range. It must be <= "end". Received {start: ${this.start}, end: ${this.end}}`

It just seems better to read for me. It is also the way how it is done in e.g. the `buffer` module.

But that is just a suggestion.

ryzokuken replied:
Hmm, I'm not as inclined regarding this one.

Let's assume that the passed value of `end` is 4, and `start` is 6. The error message would be: `The value of "start" is out of range. It must be <= ${4}. Received ${2}`.

I bet there would be a section of developers, already coping with a huge language barrier, who would might to the conclusion that there's some magic value (4 in this case) which "start" needs to be smaller than.

What about: `The value of "start" is out of range. It must be <= "end" (here: 4). Received 6` instead?

BridgeAR replied:
:+1: that works as well.

ryzokuken commented:
@BridgeAR ping.

ryzokuken commented:
@BridgeAR PTAL.

BridgeAR commented on lib/fs.js:
```diff
@@ -2014,30 +2014,40 @@ function ReadStream(path, options) {
-      const errVal = `{start: ${this.start}, end: ${this.end}}`;
-      throw new ERR_OUT_OF_RANGE('start', '<= "end"', errVal);
+    if (this.start < 0) {
+      throw new ERR_OUT_OF_RANGE('start', '=> 0 and <= 2 ** 53 - 1', this.start);
```
You were right about changing => to >= below but that should also be done here :-)

ryzokuken replied:
Good catch! I'm sorry to have missed that out, changing ASAP.

ryzokuken replied:
Ligatures to the rescue!
<img width="898" alt="screen shot 2018-05-09 at 1 22 58 am" src="https://user-images.githubusercontent.com/13121330/39779680-905bbca2-5327-11e8-9509-d609205a4960.png">

BridgeAR commented on lib/fs.js:
```diff
@@ -2014,30 +2014,40 @@ function ReadStream(path, options) {
+
+    if (this.start !== undefined && this.start > this.end) {
+      const errVal = `{start: ${this.start}, end: ${this.end}}`;
+      throw new ERR_OUT_OF_RANGE('start', `<= "end" (here: ${this.end})`, errVal);
```
It should now be `this.start` instead of `errVal`.

ryzokuken replied:
I'll remove the `errVal` variable complete, then.

ryzokuken commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14720/

ryzokuken commented:
@BridgeAR I suppose I should also try running benchmark CI and CITGM? I have no idea how to do that, though, so please help me out with that if possible.

ryzokuken commented:
Very strange that the linter says: `message: Line 2039 exceeds the maximum line length of 80.`

My editor shows Line 2039 to be 78 characters long, plus it (the line length) never even changed, it just went from `=>` to `>=`.

```console
~/Code/nodejs/node feature/create-read-stream
❯ echo "      throw new ERR_OUT_OF_RANGE('end', '>= 0 and <= 2 ** 53 - 1', this.end);" | wc -c
      78
```

ryzokuken commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14727/

ryzokuken commented:
Umm, I updated the error message? No idea how it could still fail with the same error.

```diff
-   message: 'The value of "start" is out of range. It must be <= "end" (here: 2). Received 10',
+   message: 'The value of "start" is out of range. It must be <= "end". Received {start: 10, end: 2}',
```

BridgeAR commented:
@ryzokuken it is a test that you did not touch. Since you changed the error message in general former existing tests are now failing and have to be updated as well. In this case `parallel/test-fs-read-stream-inherit`.

ryzokuken commented:
@BridgeAR sorry, I fixed this test in one file and forgot about all the others. Will do a repo-wide search and replace.

BridgeAR commented:
@ryzokuken you can check the CI results. It should only be the file that I pointed out. I recommend to run the tests on your machine and you should get the same result.

ryzokuken commented:
@BridgeAR you're right, only two files had an hardcoded check for that test and I had also updated one. Running tests in a jiffy.

ryzokuken commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14738/

ryzokuken commented:
Wait, the lint errors are still here.

ryzokuken commented:
Broke down the three lines that were failing on us, all tests *should* pass now. Waiting for tests to pass locally first this time.

ryzokuken commented:
Okay, tests pass.

CI: https://ci.nodejs.org/job/node-test-pull-request/14740/

ryzokuken commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14754/

ryzokuken commented:
Another Java error: https://ci.nodejs.org/job/node-test-commit-freebsd/17632/nodes=freebsd10-64/console

Aren't we having too many of these lately?

ryzokuken commented:
Optimistic CI run: https://ci.nodejs.org/job/node-test-pull-request/14780/

ryzokuken commented:
Rerunning: https://ci.nodejs.org/job/node-test-pull-request/14797/

ryzokuken commented:
@BridgeAR so, should I land it? Any CITGM or benchmark runs required?

BridgeAR commented:
@ryzokuken running CITGM for semver-majors is always best to detect failures early that have to be fixed. I doubt that it is going to be an issue in this case but here they are nevertheless:

CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1404/
CITGM master https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1405/

When they come out "green" (there will be quite a few failures and therefore it is easier to compare the output of those two. If they diverge, check if it has something to do with this PR) you should feel free to land the PR.

ryzokuken commented:
@BridgeAR not the exact same, but mine had a few less failures? (59 vs 66).

Also, a single less skipped test. I hope it's okay to land?

addaleax commented:
Landed in 19374fd25bac95c80e1eb877668f6b16affc36a2 :tada:
